qutebrowser/misc/mhtml.py

```diff
@@ -0,0 +1,411 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2015 Daniel Schadt
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Utils for writing a MHTML file."""
+
+import functools
+import io
+import os
+import re
+
+from collections import namedtuple
+from base64 import b64encode
+from uuid import uuid4
+from email import policy
+from email.mime.multipart import MIMEMultipart
+from email.generator import BytesGenerator
+
+from PyQt5.QtCore import QUrl
+
+from qutebrowser.utils import log, objreg, message
+
+
+_File = namedtuple("_File",
+                   "content content_type content_location transfer_encoding")
+
+
+_CSS_URL_PATTERNS = [re.compile(x) for x in [
+    rb"@import '(?P<url>[^']+)'",
+    rb'@import "(?P<url>[^"]+)"',
+    rb'''url\((?P<url>[^'"][^)]*)\)''',
+    rb'url\("(?P<url>[^"]+)"\)',
+    rb"url\('(?P<url>[^']+)'\)",
+]]
```

Comment:
"Write once, never read again" ;)

I wrote them myself, looking at the CSS docs for [@import](https://developer.mozilla.org/en-US/docs/Web/CSS/%40import) and [\<url\>](https://developer.mozilla.org/en-US/docs/Web/CSS/url).

I didn't want to include another 3rd-party-CSS-parser, and writing a new CSS parser didn't seem like the best approach either.